From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jeff Genovy <29107334+jefgen@users.noreply.github.com>
Date: Fri, 20 Nov 2020 11:35:20 -0800
Subject: ICU-PATCH: ICU-20506 Don't call UPRV_UNREACHABLE if timezoneTypes.res
 can't be loaded.


diff --git a/icu/icu4c/source/common/uloc_keytype.cpp b/icu/icu4c/source/common/uloc_keytype.cpp
index 019da058cf491dc80a2d41cdab2b496af921a032..d83ad5e1b867a78343d8de71e55218114b6dabbc 100644
--- a/icu/icu4c/source/common/uloc_keytype.cpp
+++ b/icu/icu4c/source/common/uloc_keytype.cpp
@@ -172,7 +172,14 @@ initFromResourceBundle(UErrorCode& sts) {
         LocalUResourceBundlePointer typeMapResByKey(ures_getByKey(typeMapRes.getAlias(), legacyKeyId, NULL, &tmpSts));
         if (U_FAILURE(tmpSts)) {
             // type map for each key must exist
-            UPRV_UNREACHABLE;
+
+            // ICU-20506
+            // Note: The code used to call `U_ASSERT(FALSE)` here, but this was changed in ICU-20074
+            // to unconditionally call the macro `UPRV_UNREACHABLE` instead.
+            // However, it is possible to reach this code path if 'timezoneTypes.res' can't be found or loaded.
+            // In this case the code actually _can_ carry on (it shouldn't be a fatal error), but it means that
+            // we lose the ability to convert BCP47 timezone identifiers (which aren't commonly used). 
+            U_ASSERT(FALSE);
         } else {
             LocalUResourceBundlePointer typeMapEntry;
 
