From f27726764ad4b914a96338ec8d983355442a8cfe Mon Sep 17 00:00:00 2001
From: Daniel Ju <41210545+daniel-ju@users.noreply.github.com>
Date: Tue, 4 Jun 2024 15:23:14 -0500
Subject: [PATCH] ICU-21991 added VS2022 checks and changed windows SDK version

---
 .../allinone/Build.Windows.ProjectConfiguration.props     | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/icu/icu4c/source/allinone/Build.Windows.ProjectConfiguration.props b/icu/icu4c/source/allinone/Build.Windows.ProjectConfiguration.props
index bcb0f411..5d7e3344 100644
--- a/icu/icu4c/source/allinone/Build.Windows.ProjectConfiguration.props
+++ b/icu/icu4c/source/allinone/Build.Windows.ProjectConfiguration.props
@@ -63,11 +63,11 @@
       Note: This version must match the version below in the ARM64 section for AdditionalLibraryDirectories
   -->
   <PropertyGroup Condition="'$(Platform)'=='ARM'">
-    <WindowsTargetPlatformVersion>10.0.16299.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.22621.0</WindowsTargetPlatformVersion>
     <WindowsSDKDesktopARMSupport>true</WindowsSDKDesktopARMSupport>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Platform)'=='ARM64'">
-    <WindowsTargetPlatformVersion>10.0.16299.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.22621.0</WindowsTargetPlatformVersion>
     <WindowsSDKDesktopARM64Support>true</WindowsSDKDesktopARM64Support>
   </PropertyGroup>
   <PropertyGroup>
@@ -180,7 +180,7 @@
       <!-- The ARM64 Desktop SDK doesn't include this by default -->
       <AdditionalDependencies>kernel32.lib;%(AdditionalDependencies)</AdditionalDependencies>
       <!-- Note: This needs to match the same version as WindowsTargetPlatformVersion for ARM -->
-      <AdditionalLibraryDirectories>C:\Program Files (x86)\Windows Kits\10\Lib\10.0.16299.0\um\arm</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>C:\Program Files (x86)\Windows Kits\10\Lib\10.0.22621.0\um\arm</AdditionalLibraryDirectories>
     </Link>
   </ItemDefinitionGroup>
   <!-- Options that are common to all ARM 64-bit configurations for *all* projects. -->
@@ -196,7 +196,7 @@
       <!-- The ARM64 Desktop SDK doesn't include this by default -->
       <AdditionalDependencies>kernel32.lib;%(AdditionalDependencies)</AdditionalDependencies>
       <!-- Note: This needs to match the same version as WindowsTargetPlatformVersion for ARM64 -->
-      <AdditionalLibraryDirectories>C:\Program Files (x86)\Windows Kits\10\Lib\10.0.16299.0\um\arm64</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>C:\Program Files (x86)\Windows Kits\10\Lib\10.0.22621.0\um\arm64</AdditionalLibraryDirectories>
     </Link>
   </ItemDefinitionGroup>
 </Project>
\ No newline at end of file
-- 
2.45.0.vfs.0.0

